streams/src/test/java/org/apache/kafka/streams/integration/MetricsIntegrationTest.java

```diff
@@ -0,0 +1,479 @@
+    private static final String TASK_CREATED_RATE = "task-created-rate";
+    private static final String TASK_CREATED_TOTAL = "task-created-total";
+    private static final String TASK_CLOSED_RATE = "task-closed-rate";
+    private static final String TASK_CLOSED_TOTAL = "task-closed-total";
+    private static final String SKIPPED_RECORDS_RATE = "skipped-records-rate";
+    private static final String SKIPPED_RECORDS_TOTAL = "skipped-records-total";
+    private static final String RECORD_LATENESS_AVG = "record-lateness-avg";
+    private static final String RECORD_LATENESS_MAX = "record-lateness-max";
+    private static final String HIT_RATIO_AVG = "hitRatio-avg";
+    private static final String HIT_RATIO_MIN = "hitRatio-min";
+    private static final String HIT_RATIO_MAX = "hitRatio-max";
+    private static final String TIME_WINDOWED_AGGREGATED_STREAM_STORE = "time-windowed-aggregated-stream-store";
+    private static final String SESSION_AGGREGATED_STREAM_STORE = "session-aggregated-stream-store";
+
+    private StreamsBuilder builder;
+    private Properties streamsConfiguration;
+    private KafkaStreams kafkaStreams;
+
+    // topic names
+    private static final String STREAM_INPUT = "STREAM_INPUT";
+    private static final String STREAM_OUTPUT_1 = "STREAM_OUTPUT_1";
+    private static final String STREAM_OUTPUT_2 = "STREAM_OUTPUT_2";
+    private static final String STREAM_OUTPUT_3 = "STREAM_OUTPUT_3";
+    private static final String STREAM_OUTPUT_4 = "STREAM_OUTPUT_4";
+
+    private KStream<Integer, String> stream;
+    private KStream<Integer, String> stream2;
+
+    private final String appId = "stream-metrics-test";
+
+    @Before
+    public void before() throws InterruptedException {
+        builder = new StreamsBuilder();
+        CLUSTER.createTopics(STREAM_INPUT, STREAM_OUTPUT_1, STREAM_OUTPUT_2, STREAM_OUTPUT_3, STREAM_OUTPUT_4);
+        streamsConfiguration = new Properties();
+        streamsConfiguration.put(StreamsConfig.APPLICATION_ID_CONFIG, appId);
+        streamsConfiguration.put(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, CLUSTER.bootstrapServers());
+        streamsConfiguration.put(StreamsConfig.DEFAULT_KEY_SERDE_CLASS_CONFIG, Serdes.Integer().getClass());
+        streamsConfiguration.put(StreamsConfig.DEFAULT_VALUE_SERDE_CLASS_CONFIG, Serdes.String().getClass());
+        streamsConfiguration.put(StreamsConfig.METRICS_RECORDING_LEVEL_CONFIG, Sensor.RecordingLevel.DEBUG.name);
+        streamsConfiguration.put(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG, 10 * 1024 * 1024L);
+    }
+
+    @Before
+    public void after() throws InterruptedException {
+        CLUSTER.deleteTopics(STREAM_INPUT, STREAM_OUTPUT_1, STREAM_OUTPUT_2, STREAM_OUTPUT_3, STREAM_OUTPUT_4);
+    }
+
+    private void startApplication() throws Exception {
+
+        kafkaStreams = new KafkaStreams(builder.build(), streamsConfiguration);
+        kafkaStreams.start();
+
+        Thread.sleep(10000);
```

Comment:
We avoid using time-based operations in integrations since they usually leads to flaky tests.

Consider using `TestUtils.waitForCondition()` with a timeout.

Reply:
@guozhangwang the issue is that i have no condition to be sure that the metrics has be registred to begin the test, i just tested with 5 sec (no all metric are registred) then tested with 10 sec and the test never fails. any suggestions ?

Reply:
What I was suggesting is that we can 1) remove the sleep in `startApplication` function, and 2) in the test itself, use `TestUtils.waitForCondition()` after `startApplication` for each `testXX` check, such that if it is not yet satisfied, instead of failing the test immediately, it will backoff a bit and re-execute the check and see if it is good now -- so we are giving the check continuously than doing a one-time shot.

Of course, this requires you to refactor the assertions in `testMetricByName` a bit to return booleans plus propagating the error message bottom-up so that it can shows which exact metric name did not find when we've exhausted the wait time and finally decide to fail (otherwise it will just say "fail to meet condition" but the internal information like which test function's which metric name caused it, will be lost).

Reply:
yes, thank you.